linkerd/app/outbound/src/http/peer_proxy_errors.rs

```diff
@@ -41,146 +37,89 @@ where
-        assert_eq!(response.status(), StatusCode::BAD_GATEWAY);
-        let message = response
-            .headers()
-            .get(L5D_PROXY_ERROR)
-            .expect("response did not contain l5d-proxy-error header");
-        assert_eq!(message, "proxy received invalid response");
-
-        // The client's background task should be completed without dropping
-        // the client because the connection was closed after encountering a
-        // response that contains the l5d-proxy-error header.
-        let _ = bg.await;
-    }
-
-    fn build_outbound<I>(
-        config: Config,
-        rt: ProxyRuntime,
-        connect: Connect<Endpoint>,
-    ) -> BoxNewService<Endpoint, BoxService<I, (), Error>>
-    where
-        I: io::AsyncRead + io::AsyncWrite + io::PeerAddr + Send + Unpin + 'static,
-    {
-        Outbound::new(config.clone(), rt.clone())
-            .with_stack(connect)
-            .push_http_endpoint()
-            .push_http_server()
-            .map_stack(|_, _, s| s.push_on_response(BoxRequest::layer()))
-            .push(NewServeHttp::layer(
-                config.proxy.server.h2_settings,
-                rt.drain,
-            ))
-            .map_stack(|_, _, s| s.push_on_response(BoxService::layer()))
-            .push(BoxNewService::layer())
-            .into_inner()
-    }
-
-    fn serve() -> io::Result<io::BoxedIo> {
-        let (client_io, server_io) = io::duplex(4096);
-        let http = Http::new();
-        let service = service::service_fn(move |_: Request<Body>| {
-            let response = Response::builder()
-                .status(StatusCode::BAD_GATEWAY)
-                .header(L5D_PROXY_ERROR, "proxy received invalid response")
-                .body(hyper::Body::default())
-                .unwrap();
-            future::ok::<_, Infallible>(response)
-        });
-        tokio::spawn(http.serve_connection(server_io, service));
-        Ok(io::BoxedIo::new(client_io))
+        let closed = test_util::set_client_handle(&mut req, ([192, 0, 2, 3], 50000).into());
+
+        const ERROR_MSG: &str = "something bad happened";
+        let svc = PeerProxyErrors::new(svc::mk(move |_: http::Request<hyper::Body>| {
+            future::ok::<_, Infallible>(
+                http::Response::builder()
+                    .status(http::StatusCode::BAD_GATEWAY)
+                    .header(L5D_PROXY_ERROR, ERROR_MSG)
+                    .body(hyper::Body::default())
+                    .unwrap(),
+            )
+        }));
+
+        let rsp = svc.oneshot(req).await.expect("request must succeed");
+        assert_eq!(rsp.status(), http::StatusCode::BAD_GATEWAY);
+        assert_eq!(
+            rsp.headers()
+                .get(L5D_PROXY_ERROR)
+                .expect("response did not contain l5d-proxy-error header"),
+            ERROR_MSG
+        );
+
+        // The client handle close future should fire.
+        time::timeout(time::Duration::from_secs(10), closed)
```

Comment:
It's important to use timeouts when the default (error) behavior would be to hang indefinitely

Reply:
Using `closed` to test that the client handle is closed instead of going through the full setup of a client and having to satisfy the type constraints for `http_util::connect_and_accept` is much simpler; good call.